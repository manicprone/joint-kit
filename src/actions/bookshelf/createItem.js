import objectUtils from '../../utils/object-utils'
import * as StatusErrors from '../../core/errors/status-errors'
import * as AuthUtils from '../../core/authorization/auth-utils'
import INSTANCE from '../../core/constants/instance-constants'
import ACTION from '../../core/constants/action-constants'
import * as ActionUtils from '../action-utils'
import { handleDataResponse, handleErrorResponse } from './handlers/response-handlers'

const debug = false

export default async function createItem (joint, spec = {}, input = {}, output) {
  const bookshelf = joint[INSTANCE.PROP_SERVICE]
  const trx = input[ACTION.INPUT_TRANSACTING]

  // Continue on existing transaction...
  if (trx) return performCreateItem(joint, spec, input, output)

  // Otherwise, start new transaction...
  return bookshelf.transaction((newTrx) => {
    const newInput = Object.assign({}, input)
    newInput[ACTION.INPUT_TRANSACTING] = newTrx
    return performCreateItem(joint, spec, newInput, output)
  })
}

async function performCreateItem (joint, spec = {}, input = {}, output) {
  const bookshelf = joint[INSTANCE.PROP_SERVICE]
  const modelName = spec[ACTION.SPEC_MODEL_NAME]
  const specFields = spec[ACTION.SPEC_FIELDS]
  const specAuth = spec[ACTION.SPEC_AUTH] || {}
  const authRules = specAuth[ACTION.SPEC_AUTH_RULES]
  const inputFields = ActionUtils.prepareFieldData(specFields, input[ACTION.INPUT_FIELDS])
  const trx = input[ACTION.INPUT_TRANSACTING]
  const authContext = input[ACTION.INPUT_AUTH_CONTEXT]

  // Reject if model does not exist...
  const model = bookshelf.model(modelName)
  if (!model) {
    if (debug) console.log(`[JOINT] [action:createItem] The model "${modelName}" is not recognized`)
    return Promise.reject(StatusErrors.generateModelNotRecognizedError(modelName))
  }

  // Reject when required fields are not provided...
  const requiredFieldCheck = ActionUtils.checkRequiredFields(specFields, inputFields)
  if (!requiredFieldCheck.satisfied) {
    if (debug) console.log('[JOINT] [action:createItem] Action has missing required fields:', requiredFieldCheck.missing)
    return Promise.reject(StatusErrors.generateMissingFieldsError(requiredFieldCheck.missing))
  }

  // Respect auth...
  if (authRules) {
    const ownerCreds = ActionUtils.parseOwnerCreds(specAuth, inputFields)
    if (!AuthUtils.isAllowed(authContext, authRules, ownerCreds)) {
      return Promise.reject(StatusErrors.generateNotAuthorizedError())
    }
  } // end-if (authRules)

  // Prepare action options...
  const actionOpts = {}
  if (trx) actionOpts.transacting = trx

  // Build create package...
  const createData = {}
  if (inputFields && specFields) {
    specFields.forEach((fieldSpec) => {
      const fieldName = fieldSpec.name
      const hasDefault = objectUtils.has(fieldSpec, ACTION.SPEC_FIELDS_OPT_DEFAULT_VALUE)
      const defaultValue = (hasDefault) ? ActionUtils.processDefaultValue(inputFields, fieldSpec[ACTION.SPEC_FIELDS_OPT_DEFAULT_VALUE]) : null
      const hasInput = objectUtils.has(inputFields, fieldName)
      const isLocked = objectUtils.get(fieldSpec, ACTION.SPEC_FIELDS_OPT_LOCKED, false)

      if (hasInput || hasDefault) {
        const inputValue = !isLocked && hasInput ? inputFields[fieldName].value : defaultValue
        createData[fieldName] = inputValue
      }
    }) // end-specFields.forEach
  } // end-if (inputFields && specFields)

  // Debug executing logic...
  if (debug) console.log(`[JOINT] [action:createItem] EXECUTING => CREATE ${modelName} WITH`, createData)

  try {
    // Create item...
    const data = await model.forge(createData).save(null, actionOpts)
    return handleDataResponse(joint, modelName, data, output)
  } catch (error) {
    return handleErrorResponse(error, 'createItem', modelName)
  }
} // END - performCreateItem
