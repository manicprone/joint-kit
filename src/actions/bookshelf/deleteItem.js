import objectUtils from '../../utils/object-utils'
import * as StatusErrors from '../../core/errors/status-errors'
import * as AuthUtils from '../../core/authorization/auth-utils'
import INSTANCE from '../../core/constants/instance-constants'
import ACTION from '../../core/constants/action-constants'
import * as ActionUtils from '../action-utils'
import * as BookshelfUtils from './utils/bookshelf-utils'
import { handleDataResponse, handleErrorResponse } from './handlers/response-handlers'

const debug = false

export default async function deleteItem (joint, spec = {}, input = {}, output) {
  const bookshelf = joint[INSTANCE.PROP_SERVICE]
  const trx = input[ACTION.INPUT_TRANSACTING]

  // Continue on existing transaction...
  if (trx) return performDeleteItem(joint, spec, input, output)

  // Otherwise, start new transaction...
  return bookshelf.transaction((newTrx) => {
    const newInput = Object.assign({}, input)
    newInput[ACTION.INPUT_TRANSACTING] = newTrx
    return performDeleteItem(joint, spec, newInput, output)
  })
}

async function performDeleteItem (joint, spec = {}, input = {}, output) {
  const bookshelf = joint[INSTANCE.PROP_SERVICE]
  const modelName = spec[ACTION.SPEC_MODEL_NAME]
  const specFields = ActionUtils.normalizeFieldSpec(spec[ACTION.SPEC_FIELDS])
  const specAuth = spec[ACTION.SPEC_AUTH] || {}
  const inputFields = ActionUtils.prepareFieldData(specFields, input[ACTION.INPUT_FIELDS])
  const trx = input[ACTION.INPUT_TRANSACTING]
  const authContext = input[ACTION.INPUT_AUTH_CONTEXT]

  // Reject if model does not exist...
  const model = bookshelf.model(modelName)
  if (!model) {
    if (debug) console.log(`[JOINT] [action:deleteItem] The model "${modelName}" is not recognized`)
    return Promise.reject(StatusErrors.generateModelNotRecognizedError(modelName))
  }

  // Reject when required fields are not provided...
  const requiredFieldCheck = ActionUtils.checkRequiredFields(specFields, inputFields)
  if (!requiredFieldCheck.satisfied) {
    if (debug) console.log('[JOINT] [action:deleteItem] Action has missing required fields:', requiredFieldCheck.missing)
    return Promise.reject(StatusErrors.generateMissingFieldsError(requiredFieldCheck.missing))
  }

  // Perform lookup of item first, if specified...
  const lookupFieldData = ActionUtils.getLookupFieldData(specFields, inputFields)
  if (lookupFieldData) {
    return doLookupThenAction(joint, lookupFieldData, modelName, specFields, specAuth, inputFields, authContext, trx, output)
  }

  // Otherwise, just perform action...
  return doAction(joint, modelName, specFields, specAuth, null, inputFields, authContext, trx, output)
} // END - performDeleteItem

async function doLookupThenAction (joint, lookupFieldData, modelName, specFields, specAuth, inputFields, authContext, trx, output) {
  const bookshelf = joint[INSTANCE.PROP_SERVICE]
  const getItemOpts = { require: true }
  if (trx) getItemOpts.transacting = trx

  try {
    const resource = await bookshelf.model(modelName).query((queryBuilder) => {
      Object.entries(lookupFieldData)
        .forEach(([fieldName, field]) => {
          BookshelfUtils.appendWhereClause(queryBuilder, fieldName, field.value, field.matchStrategy)
        })
    }).fetch(getItemOpts)

    // Prepare ownerCreds from retrieved data...
    const combinedFields = Object.assign({}, resource.attributes, inputFields)
    const ownerCreds = ActionUtils.parseOwnerCreds(specAuth, combinedFields)

    return doAction(joint, modelName, specFields, specAuth, ownerCreds, inputFields, authContext, trx, output)
  } catch (error) {
    return handleErrorResponse(error, 'deleteItem', modelName)
  }
} // END - doLookupThenAction

async function doAction (joint, modelName, specFields, specAuth, ownerCreds, inputFields, authContext, trx, output) {
  const bookshelf = joint[INSTANCE.PROP_SERVICE]

  // Respect auth...
  const authRules = specAuth[ACTION.SPEC_AUTH_RULES]
  if (authRules) {
    const ownerCredsData = ownerCreds || ActionUtils.parseOwnerCreds(specAuth, inputFields)
    if (!AuthUtils.isAllowed(authContext, authRules, ownerCredsData)) {
      return Promise.reject(StatusErrors.generateNotAuthorizedError())
    }
  } // end-if (authRules)

  // Prepare action options...
  const actionOpts = { require: true }
  if (trx) actionOpts.transacting = trx

  // Build where clause...
  const queryOpts = (queryBuilder) => {
    if (inputFields && specFields) {
      specFields.forEach((fieldSpec) => {
        const fieldName = fieldSpec.name
        const hasInput = objectUtils.has(inputFields, fieldName)
        const matchStrategy = objectUtils.get(inputFields,
          `${fieldName}.matchStrategy`,
          ACTION.INPUT_FIELD_MATCHING_STRATEGY_EXACT
        )
        if (hasInput) {
          BookshelfUtils.appendWhereClause(queryBuilder, fieldName, inputFields[fieldName].value, matchStrategy)
        }
      })
    } // end-if (inputFields && specFields)
  }

  // Debug executing logic...
  if (debug) console.log(`[JOINT] [action:deleteItem] EXECUTING => DELETE ${modelName} WHERE`, queryOpts)

  try {
    // Delete item...
    const data = await bookshelf.model(modelName).query(queryOpts).destroy(actionOpts)
    return handleDataResponse(joint, modelName, data, output)
  } catch (error) {
    return handleErrorResponse(error, 'deleteItem', modelName)
  }
} // END - doAction
