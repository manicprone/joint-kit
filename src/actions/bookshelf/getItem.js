import objectUtils from '../../utils/object-utils'
import * as StatusErrors from '../../core/errors/status-errors'
import * as AuthUtils from '../../core/authorization/auth-utils'
import INSTANCE from '../../core/constants/instance-constants'
import ACTION from '../../core/constants/action-constants'
import * as ActionUtils from '../action-utils'
import * as BookshelfUtils from './utils/bookshelf-utils'
import { handleDataResponse, handleErrorResponse } from './handlers/response-handlers'

const debug = false

export default async function getItem (joint, spec = {}, input = {}, output) {
  const bookshelf = joint[INSTANCE.PROP_SERVICE]
  const modelName = spec[ACTION.SPEC_MODEL_NAME]
  const specFields = ActionUtils.normalizeFieldSpec(spec[ACTION.SPEC_FIELDS])
  const specAuth = spec[ACTION.SPEC_AUTH] || {}
  const authRules = specAuth[ACTION.SPEC_AUTH_RULES]
  const returnColsDef = spec[ACTION.SPEC_FIELDS_TO_RETURN]
  const inputFields = ActionUtils.prepareFieldData(specFields, input[ACTION.INPUT_FIELDS])
  const trx = input[ACTION.INPUT_TRANSACTING]
  const authContext = input[ACTION.INPUT_AUTH_CONTEXT]

  // Reject if model does not exist...
  const model = bookshelf.model(modelName)
  if (!model) {
    if (debug) console.log(`[JOINT] [action:getItem] The model "${modelName}" is not recognized`)
    return Promise.reject(StatusErrors.generateModelNotRecognizedError(modelName))
  }

  // Reject when required fields are not provided...
  const requiredFieldCheck = ActionUtils.checkRequiredFields(specFields, inputFields)
  if (!requiredFieldCheck.satisfied) {
    if (debug) console.log('[JOINT] [action:getItem] Action has missing required fields:', requiredFieldCheck.missing)
    return Promise.reject(StatusErrors.generateMissingFieldsError(requiredFieldCheck.missing))
  }

  // Prepare action options...
  const actionOpts = { require: true }
  if (trx) actionOpts.transacting = trx

  // Restrict columns to return with payload...
  if (returnColsDef) {
    // If a single set (array) is defined, honor the setting...
    if (Array.isArray(returnColsDef)) {
      actionOpts.columns = returnColsDef

    // Otherwise, try to honor the set requested by the input...
    } else if (input[ACTION.INPUT_FIELD_SET] && objectUtils.has(returnColsDef, input[ACTION.INPUT_FIELD_SET])) {
      actionOpts.columns = returnColsDef[input[ACTION.INPUT_FIELD_SET]]

    // If the input does not declare a set, check for a "default" set...
    } else if (returnColsDef.default && Array.isArray(returnColsDef.default)) {
      actionOpts.columns = returnColsDef.default
    }
  } // end-if (returnColsDef)

  // -------------------------------------------------------------------------
  // Include associations (associations & loadDirect will be combined)
  // -------------------------------------------------------------------------
  // Handle "loadDirect" option...
  const inputLoadDirectDef = input[ACTION.INPUT_LOAD_DIRECT] || []
  const loadDirectDef = (spec[ACTION.SPEC_FORCE_LOAD_DIRECT])
    ? spec[ACTION.SPEC_FORCE_LOAD_DIRECT].concat(inputLoadDirectDef)
    : inputLoadDirectDef
  const loadDirect = ActionUtils.parseLoadDirect(loadDirectDef)
  // Handle "associations" option...
  const inputAssocs = input[ACTION.INPUT_ASSOCIATIONS] || []
  const assocs = (spec[ACTION.SPEC_FORCE_ASSOCIATIONS])
    ? objectUtils.union(spec[ACTION.SPEC_FORCE_ASSOCIATIONS], inputAssocs)
    : inputAssocs
  // Combine...
  const allAssociations = (loadDirect.associations && loadDirect.associations.length > 0)
    ? objectUtils.union(assocs, loadDirect.associations)
    : assocs
  if (allAssociations.length > 0) actionOpts.withRelated = allAssociations

  // Prepare query...
  const queryOpts = (queryBuilder) => {
    if (inputFields && specFields) {
      specFields.forEach((fieldSpec) => {
        const fieldName = fieldSpec.name
        const hasDefault = objectUtils.has(fieldSpec, ACTION.SPEC_FIELDS_OPT_DEFAULT_VALUE)
        const defaultValue = (hasDefault) ? ActionUtils.processDefaultValue(inputFields, fieldSpec[ACTION.SPEC_FIELDS_OPT_DEFAULT_VALUE]) : null
        const hasInput = objectUtils.has(inputFields, fieldName)
        const matchStrategy = objectUtils.get(inputFields,
          `${fieldName}.matchStrategy`,
          ACTION.INPUT_FIELD_MATCHING_STRATEGY_EXACT
        )
        const isLocked = objectUtils.get(fieldSpec, ACTION.SPEC_FIELDS_OPT_LOCKED, false)

        if (!isLocked && (hasInput || hasDefault)) {
          const inputValue = (hasInput)
            ? inputFields[fieldName].value
            : defaultValue
          BookshelfUtils.appendWhereClause(queryBuilder, fieldName, inputValue, matchStrategy)
        } else if (isLocked && hasDefault) {
          BookshelfUtils.appendWhereClause(queryBuilder, fieldName, defaultValue, matchStrategy)
        }
      }) // end-specFields.forEach
    } // end-if (inputFields && specFields)
  }

  // Debug executing logic...
  if (debug) console.log(`[JOINT] [action:getItem] EXECUTING => GET ${modelName} QUERY`, queryOpts)

  try {
    // Get item...
    const data = await model.query(queryOpts).fetch(actionOpts)

    // Respect auth...
    if (authRules) {
      const combinedFields = Object.assign({}, data.attributes, inputFields)
      const ownerCreds = ActionUtils.parseOwnerCreds(specAuth, combinedFields)
      if (!AuthUtils.isAllowed(authContext, authRules, ownerCreds)) {
        return Promise.reject(StatusErrors.generateNotAuthorizedError())
      }
    } // end-if (authRules)

    // Handle loadDirect requests...
    if (loadDirect.associations) BookshelfUtils.loadRelationsToItemBase(data, loadDirect, input.associations)

    // Return data...
    return handleDataResponse(joint, modelName, data, output)
  } catch (error) {
    return handleErrorResponse(error, 'getItem', modelName, inputAssocs)
  }
} // END - getItem
