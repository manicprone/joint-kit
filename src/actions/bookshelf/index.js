// ---------------
// Base actions...
// ---------------
export { default as createItem } from './createItem'
export { default as upsertItem } from './upsertItem'
export { default as updateItem } from './updateItem'
export { default as updateMany } from './updateMany'
export { default as getItem } from './getItem'
export { default as getItems } from './getItems'
export { default as deleteItem } from './deleteItem'

// ----------------------
// Association actions...
// ----------------------
export { default as addAssociatedItems } from './addAssociatedItems'
export { default as hasAssociatedItem } from './hasAssociatedItem'
export { default as getAllAssociatedItems } from './getAllAssociatedItems'
export { default as removeAssociatedItems } from './removeAssociatedItems'
export { default as removeAllAssociatedItems } from './removeAllAssociatedItems'
