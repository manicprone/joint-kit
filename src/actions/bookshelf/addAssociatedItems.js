import * as StatusErrors from '../../core/errors/status-errors'
import INSTANCE from '../../core/constants/instance-constants'
import ACTION from '../../core/constants/action-constants'
import getItem from './getItem'
import getItems from './getItems'
import { handleDataResponse, handleErrorResponse } from './handlers/response-handlers'

const debug = false

export default async function addAssociatedItems(joint, spec = {}, input = {}, output) {
  const bookshelf = joint[INSTANCE.PROP_SERVICE]
  const trx = input[ACTION.INPUT_TRANSACTING]

  // Continue on existing transaction...
  if (trx) return performAddAssociatedItems(joint, spec, input, output)

  // Otherwise, start new transaction...
  return bookshelf.transaction((newTrx) => {
    const newInput = Object.assign({}, input)
    newInput[ACTION.INPUT_TRANSACTING] = newTrx
    return performAddAssociatedItems(joint, spec, newInput, output)
  })
}

async function performAddAssociatedItems(joint, spec = {}, input = {}, output) {
  const specMain = spec[ACTION.RESOURCE_MAIN]
  const modelNameMain = (specMain) ? specMain[ACTION.SPEC_MODEL_NAME] : null
  const specAssoc = spec[ACTION.RESOURCE_ASSOCIATION]
  const assocName = (specAssoc) ? specAssoc[ACTION.SPEC_ASSOCIATION_NAME] : null
  const inputMain = input[ACTION.RESOURCE_MAIN]
  const inputAssoc = input[ACTION.RESOURCE_ASSOCIATION]
  const trx = input[ACTION.INPUT_TRANSACTING]

  // Reject when required properties are not provided...
  const missingProps = []
  if (!specMain) missingProps.push(`spec.${ACTION.RESOURCE_MAIN}`)
  if (!specAssoc) missingProps.push(`spec.${ACTION.RESOURCE_ASSOCIATION}`)
  if (!assocName) missingProps.push(`spec.${ACTION.RESOURCE_ASSOCIATION}.${ACTION.SPEC_ASSOCIATION_NAME}`)
  if (!inputMain) missingProps.push(`input.${ACTION.RESOURCE_MAIN}`)
  if (!inputAssoc) missingProps.push(`input.${ACTION.RESOURCE_ASSOCIATION}`)
  if (missingProps.length > 0) {
    if (debug) console.log(`[JOINT] [action:addAssociatedItems] Required properties missing: "${missingProps.join('", "')}"`)
    return Promise.reject(StatusErrors.generateInvalidAssociationPropertiesError(missingProps))
  }

  // Lookup model name of association, add to spec if not provided...
  let modelNameAssoc = specAssoc[ACTION.SPEC_MODEL_NAME]
  if (!modelNameAssoc) {
    modelNameAssoc = (joint.modelNameOfAssoc[modelNameMain]) ? joint.modelNameOfAssoc[modelNameMain][assocName] : null
    specAssoc[ACTION.SPEC_MODEL_NAME] = modelNameAssoc
  }

  // Load trx to both resources...
  inputMain[ACTION.INPUT_TRANSACTING] = trx
  inputAssoc[ACTION.INPUT_TRANSACTING] = trx

  // Include existing associations of this type (for return)...
  inputMain[ACTION.INPUT_ASSOCIATIONS] = [assocName]

  try {
    // Lookup resources...
    const main = await getItem(joint, specMain, inputMain)
    const assoc = await getItems(joint, specAssoc, inputAssoc)

    // TODO: It looks like the latest version of Bookshelf no longer guards against duplicate assoc (via attach) !!!
    //       We will have to provide the logic here !!!
    // const existingAssoc = main.relations[assocName]
    // console.log('[DEV] existing assocs =>', existingAssoc.models)
    // console.log('[DEV] assocs to add =>', assoc.models)

    // Reject with 404 if instances of the requested association were not found...
    if (assoc.length === 0) {
      return Promise.reject(StatusErrors.generateAssociatedItemsDoNotExistError(modelNameAssoc))
    }

    // Otherwise, attach associations to main...
    await main.related(assocName).attach(assoc.models, { transacting: trx })
    return handleDataResponse(joint, modelNameMain, main, output)

  } catch (error) {
    return handleErrorResponse(error, 'addAssociatedItems', modelNameMain)
  }
} // END - performAddAssociatedItems
