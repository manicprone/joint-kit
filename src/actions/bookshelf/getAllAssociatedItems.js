import * as StatusErrors from '../../core/errors/status-errors'
import INSTANCE from '../../core/constants/instance-constants'
import ACTION from '../../core/constants/action-constants'
import getItem from './getItem'
import { handleDataResponse, handleErrorResponse } from './handlers/response-handlers'

const debug = false

export default async function getAllAssociatedItems(joint, spec = {}, input = {}, output) {
  const bookshelf = joint[INSTANCE.PROP_SERVICE]
  const trx = input[ACTION.INPUT_TRANSACTING]

  // Continue on existing transaction...
  if (trx) return performGetAllAssociatedItems(joint, spec, input, output)

  // Otherwise, start new transaction...
  return bookshelf.transaction((newTrx) => {
    const newInput = Object.assign({}, input)
    newInput[ACTION.INPUT_TRANSACTING] = newTrx
    return performGetAllAssociatedItems(joint, spec, newInput, output)
  })
}

async function performGetAllAssociatedItems(joint, spec = {}, input = {}, output) {
  const specMain = spec[ACTION.RESOURCE_MAIN]
  const modelNameMain = (specMain) ? specMain[ACTION.SPEC_MODEL_NAME] : null
  const specAssoc = spec[ACTION.RESOURCE_ASSOCIATION]
  const assocName = (specAssoc) ? specAssoc[ACTION.SPEC_ASSOCIATION_NAME] : null
  const inputMain = input[ACTION.RESOURCE_MAIN] ? { ...input[ACTION.RESOURCE_MAIN] } : null
  const trx = input[ACTION.INPUT_TRANSACTING]

  // Reject when required properties are not provided...
  const missingProps = []
  if (!specMain) missingProps.push(`spec.${ACTION.RESOURCE_MAIN}`)
  if (!specAssoc) missingProps.push(`spec.${ACTION.RESOURCE_ASSOCIATION}`)
  if (!assocName) missingProps.push(`spec.${ACTION.RESOURCE_ASSOCIATION}.${ACTION.SPEC_ASSOCIATION_NAME}`)
  if (!inputMain) missingProps.push(`input.${ACTION.RESOURCE_MAIN}`)
  if (missingProps.length > 0) {
    if (debug) console.log(`[JOINT] [action:getAllAssociatedItems] Required properties missing: "${missingProps.join('", "')}"`)
    return Promise.reject(StatusErrors.generateInvalidAssociationPropertiesError(missingProps))
  }

  // Lookup model name of association, add to spec if not provided...
  let modelNameAssoc = specAssoc[ACTION.SPEC_MODEL_NAME]
  if (!modelNameAssoc) {
    modelNameAssoc = (joint.modelNameOfAssoc[modelNameMain]) ? joint.modelNameOfAssoc[modelNameMain][assocName] : null
    specAssoc[ACTION.SPEC_MODEL_NAME] = modelNameAssoc
  }

  // Load trx to main resource...
  inputMain[ACTION.INPUT_TRANSACTING] = trx

  // Return existing associations of this type...
  inputMain[ACTION.INPUT_ASSOCIATIONS] = [assocName]

  try {
    // Lookup main resource...
    const main = await getItem(joint, specMain, inputMain)

    // Access all items of the requested association...
    const assoc = main.related(assocName)

    // Reject with 404 if instances of the requested association were not found...
    if (assoc.length === 0) {
      return Promise.reject(StatusErrors.generateAssociatedItemsDoNotExistError(modelNameAssoc))
    }

    // Otherwise, return associated items...
    return handleDataResponse(joint, modelNameAssoc, assoc, output)

  } catch (error) {
    return handleErrorResponse(error, 'getAllAssociatedItems', modelNameAssoc)
  }
} // END - performGetAllAssociatedItems
