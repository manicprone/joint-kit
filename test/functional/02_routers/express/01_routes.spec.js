import chai, { expect } from 'chai'
import chaiHttp from 'chai-http'
import chaiAsPromised from 'chai-as-promised'
import express from 'express'
import Joint from '../../../../src'
import blogAppModels from '../../../scenarios/blog-app/model-config'
import blogAppMethods from '../../../scenarios/blog-app/method-config'
import blogAppRoutes from '../../../scenarios/blog-app/route-config'
import bookshelf from '../../../db/bookshelf/service'
import { resetDB } from '../../../db/bookshelf/db-utils'

chai.use(chaiHttp)
chai.use(chaiAsPromised)

const apiBasePath = '/api/express'
const apiPort = 9999
const apiURL = `http://localhost:${apiPort}${apiBasePath}`

let blogApp = null

describe.skip('CUSTOM ROUTER SIMULATION [express]', () => {
  before(() => {
    // --------
    // Blog App
    // --------
    blogApp = new Joint({ service: bookshelf, server: express, output: 'json-api' })
    blogApp.generate({
      modelConfig: blogAppModels,
      methodConfig: blogAppMethods,
      routeConfig: blogAppRoutes,
      log: false,
    })

    // Run express server with configured router...
    const app = express()
    app.use(apiBasePath, blogApp.router)
    app.listen(apiPort)
  })

  // ---------------------------------------------------------------------------
  // Resource: User
  // ---------------------------------------------------------------------------
  describe('User', () => {

    // -----------------
    // Route: GET /users
    // -----------------
    describe('GET /users', () => {
      before(() => resetDB(['users']))

      it('should return (200) with the requested collection of users', () => {
        const resourceURI = '/users'

        return chai.request(apiURL).get(resourceURI)
          .then((res) => {
            const resBody = res.body
            const data = resBody.data
            const meta = resBody.meta

            expect(res).to.have.status(200)
            expect(resBody).to.have.keys(['data', 'meta'])

            expect(data).to.have.length(10)
            expect(meta).to.have.keys(['total_items'])
            expect(meta).to.contain({
              total_items: 10,
            })

            const firstItem = data[0]
            expect(firstItem).to.have.property('attributes')
            expect(firstItem).to.contain({
              type: 'User',
              id: 10,
            })
          })
      })

      it('should support paginated requests when the "skip" and "limit" parameters are used', () => {
        const skip = 0
        const limit = 3
        const resourceURI = '/users'
        const queryString = `skip=${skip}&limit=${limit}`

        return chai.request(apiURL).get(resourceURI).query(queryString)
          .then((res) => {
            const resBody = res.body
            const data = resBody.data
            const meta = resBody.meta

            expect(res).to.have.status(200)
            expect(resBody).to.have.keys(['data', 'meta'])

            expect(data).to.have.length(limit)
            expect(meta).to.have.keys(['total_items', 'skip', 'limit'])
            expect(meta).to.contain({
              total_items: 10,
              skip,
              limit,
            })
          })
      })
    }) // END - GET /users

    // --------------------
    // Route: GET /user/:id
    // --------------------
    describe('GET /user/:id', () => {
      before(() => resetDB(['users']))

      it('should return (200) with the requested user', () => {
        const userID = 1
        const resourceURI = `/user/${userID}`

        return chai.request(apiURL).get(resourceURI)
          .then((res) => {
            const resBody = res.body
            const data = resBody.data

            expect(res).to.have.status(200)
            expect(resBody).to.have.keys(['data'])
            expect(data).to.contain({
              type: 'User',
              id: userID,
            })
            expect(data.attributes).to.contain({
              display_name: 'Supa Admin',
              username: 'super-admin',
            })
          })
      })

      it('should return associations when the "with" parameter is used', () => {
        const userID = 6
        const assocName = 'info'
        const resourceURI = `/user/${userID}`
        const queryString = `with=${assocName}`

        return chai.request(apiURL).get(resourceURI).query(queryString)
          .then((res) => {
            const resBody = res.body
            const data = resBody.data

            expect(res).to.have.status(200)
            expect(resBody).to.have.keys(['data', 'included'])
            expect(data).to.contain({
              type: 'User',
              id: userID,
            })
            expect(data.attributes).to.contain({
              username: 'ricksanchez',
            })
            expect(data.relationships).to.have.keys([assocName])
          })
      })

      it('should load association data directly to the base attributes when the "load" parameter is used', () => {
        const userID = 4
        const assocName = 'info'
        const assocField = 'professional_title'
        const resourceURI = `/user/${userID}`
        const queryString = `load=${assocName}:${assocField}`

        return chai.request(apiURL).get(resourceURI).query(queryString)
          .then((res) => {
            const resBody = res.body
            const data = resBody.data

            expect(res).to.have.status(200)
            expect(resBody).to.have.keys(['data'])

            expect(data).to.contain({
              type: 'User',
              id: userID,
            })

            expect(data.attributes)
              .to.contain({
                info: 'EdgeCaser',
              })
          })
      })
    }) // END - GET /user/:id

  }) // END - User
})
