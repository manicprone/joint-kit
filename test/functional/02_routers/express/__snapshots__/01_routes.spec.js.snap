// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CUSTOM ROUTER SIMULATION [express] > User > GET /user/:id > should load association data directly to the base attributes when the "load" parameter is used 1`] = `
{
  "data": {
    "attributes": {
      "avatar_url": null,
      "display_name": "The Manic Edge",
      "info": "EdgeCaser",
      "username": "the_manic_edge",
    },
    "id": 4,
    "type": "User",
  },
}
`;

exports[`CUSTOM ROUTER SIMULATION [express] > User > GET /user/:id > should return (200) with the requested user 1`] = `
{
  "data": {
    "attributes": {
      "avatar_url": null,
      "display_name": "Supa Admin",
      "username": "super-admin",
    },
    "id": 1,
    "type": "User",
  },
}
`;

exports[`CUSTOM ROUTER SIMULATION [express] > User > GET /user/:id > should return associations when the "with" parameter is used 1`] = `
{
  "attributes": {
    "avatar_url": null,
    "display_name": "Rick",
    "username": "ricksanchez",
  },
  "id": 6,
  "relationships": {
    "info": {
      "data": {
        "id": 3,
        "type": "UserInfo",
      },
    },
  },
  "type": "User",
}
`;

exports[`CUSTOM ROUTER SIMULATION [express] > User > GET /user/:id > should return associations when the "with" parameter is used 2`] = `
{
  "created_at": Any<String>,
  "description": null,
  "professional_title": "Rickforcer",
  "tagline": "The Rickiest",
  "updated_at": Any<String>,
  "user_id": 6,
}
`;

exports[`CUSTOM ROUTER SIMULATION [express] > User > GET /users > should return (200) with the requested collection of users 1`] = `
{
  "data": [
    {
      "attributes": {
        "avatar_url": null,
        "display_name": "Z'araq",
        "username": "zaraq",
      },
      "id": 11,
      "type": "User",
    },
    {
      "attributes": {
        "avatar_url": null,
        "display_name": "Summer",
        "username": "summersmith",
      },
      "id": 10,
      "type": "User",
    },
    {
      "attributes": {
        "avatar_url": null,
        "display_name": "Beth",
        "username": "bethsmith",
      },
      "id": 9,
      "type": "User",
    },
    {
      "attributes": {
        "avatar_url": null,
        "display_name": "Jerry",
        "username": "jerrysmith",
      },
      "id": 8,
      "type": "User",
    },
    {
      "attributes": {
        "avatar_url": null,
        "display_name": "Morty",
        "username": "mortysmith",
      },
      "id": 7,
      "type": "User",
    },
    {
      "attributes": {
        "avatar_url": null,
        "display_name": "Rick",
        "username": "ricksanchez",
      },
      "id": 6,
      "type": "User",
    },
    {
      "attributes": {
        "avatar_url": null,
        "display_name": "Segmented",
        "username": "segmented",
      },
      "id": 5,
      "type": "User",
    },
    {
      "attributes": {
        "avatar_url": null,
        "display_name": "The Manic Edge",
        "username": "the_manic_edge",
      },
      "id": 4,
      "type": "User",
    },
    {
      "attributes": {
        "avatar_url": null,
        "display_name": "Hot Mod",
        "username": "hotmod",
      },
      "id": 3,
      "type": "User",
    },
    {
      "attributes": {
        "avatar_url": null,
        "display_name": "Admin",
        "username": "admin",
      },
      "id": 2,
      "type": "User",
    },
    {
      "attributes": {
        "avatar_url": null,
        "display_name": "Supa Admin",
        "username": "super-admin",
      },
      "id": 1,
      "type": "User",
    },
  ],
  "meta": {
    "total_items": 11,
  },
}
`;

exports[`CUSTOM ROUTER SIMULATION [express] > User > GET /users > should support paginated requests when the "skip" and "limit" parameters are used 1`] = `
{
  "data": [
    {
      "attributes": {
        "avatar_url": null,
        "display_name": "Z'araq",
        "username": "zaraq",
      },
      "id": 11,
      "type": "User",
    },
    {
      "attributes": {
        "avatar_url": null,
        "display_name": "Summer",
        "username": "summersmith",
      },
      "id": 10,
      "type": "User",
    },
    {
      "attributes": {
        "avatar_url": null,
        "display_name": "Beth",
        "username": "bethsmith",
      },
      "id": 9,
      "type": "User",
    },
  ],
  "meta": {
    "limit": 3,
    "skip": 0,
    "total_items": 11,
  },
}
`;
